semantic-ast/src/AST/GenerateSyntax.hs

```diff
@@ -62,76 +66,80 @@ getAllSymbols language = do
-      tsSymbol <- runIO $ withCStringLen datatypeName (\(s, len) -> TS.ts_language_symbol_for_name language s len False)
-      pure [ TySynD name [] (ConT ''Token `AppT` LitT (StrTyLit datatypeName) `AppT` LitT (NumTyLit (fromIntegral tsSymbol))) ]
-    LeafType (DatatypeName datatypeName) Named -> do
-      con <- ctorForLeafType (DatatypeName datatypeName) typeParameterName
-      symbolMatchingInstance <- symbolMatchingInstance allSymbols name Named datatypeName
-      traversalInstances <- makeTraversalInstances (conT name)
-      pure
-        (  generatedDatatype name [con] typeParameterName
-        :  symbolMatchingInstance
-        <> traversalInstances)
+      let tsSymbol = runIO $ withCStringLen datatypeName (\(s, len) -> TS.ts_language_symbol_for_name language s len False)
+      fmap (pure @[]) (tySynD name [] (conT ''Token `appT` litT (strTyLit datatypeName) `appT` litT (tsSymbol >>= numTyLit . fromIntegral)))
+    LeafType datatypeName Named ->
+      let con = ctorForLeafType datatypeName annParameterName
+          symbols = symbolMatchingInstance allSymbols name Named datatypeName
+      in glue <$> generatedDatatype [con] <*> symbols <*> traversalInstances
   where
     -- Skip generating datatypes that have already been defined (overridden) in the module where the splice is running.
     skipDefined m = do
       isLocal <- lookupTypeName nameStr >>= maybe (pure False) isLocalName
       if isLocal then pure [] else m
-    name = mkName nameStr
     nameStr = toNameString (datatypeNameStatus datatype) (getDatatypeName (AST.Deserialize.datatypeName datatype))
-    deriveStockClause = DerivClause (Just StockStrategy) [ ConT ''Eq, ConT ''Ord, ConT ''Show, ConT ''Generic, ConT ''Generic1]
-    deriveAnyClassClause = DerivClause (Just AnyclassStrategy) [ConT ''TS.Unmarshal, ConT ''Traversable1 `AppT` VarT (mkName "someConstraint")]
-    deriveGN = DerivClause (Just NewtypeStrategy) [ConT ''TS.SymbolMatching]
-    generatedDatatype name cons typeParameterName = DataD [] name [PlainTV typeParameterName] Nothing cons [deriveStockClause, deriveAnyClassClause]
 
+makeStandaloneDerivings :: TypeQ -> Q [Dec]
+makeStandaloneDerivings ty =
+  [d|
+   deriving instance ((forall x . Eq x => Eq (f x)), Eq a) => Eq ($ty f a)
+   deriving instance ((forall x . Ord x => Ord (f x)), (forall x . Eq x => Eq (f x)), Ord a) => Ord ($ty f a)
+   deriving instance ((forall x . Show x => Show (f x)), Show a) => Show ($ty f a)
+   instance TS.Unmarshal ($ty Identity)
+
+   |]
 
 makeTraversalInstances :: TypeQ -> Q [Dec]
 makeTraversalInstances ty =
   [d|
-  instance Foldable $ty where
+  instance Traversable f => Foldable ($ty f) where
     foldMap = foldMapDefault1
-  instance Functor $ty where
+  instance Traversable f => Functor ($ty f) where
     fmap = fmapDefault1
-  instance Traversable $ty where
+  instance Traversable f => Traversable ($ty f) where
     traverse = traverseDefault1
   |]
```

Comment:
We fix the `Traversable`, `Foldable`, and `Functor` instances to account for the new kindedness.